src/Clean/Clean.man
.\" t
.\" @(#)Clean.1		02/07/00
.TH Clean 1.8 "Feb 7 2000"
.UC
.SH NAME
\fBClean\fP \- the AfterStep desktop clutter reduction module

.SH SYNOPSIS
\fBClean\fP [\-f \fIconfig_file\fP] [\-v | \--version] [\-h | \--help]
.sp
Although \fBClean\fP can be run from the command line, it is normally 
spawned by afterstep in the \fIautoexec\fP file, in the following manner:
.nf
.sp
        Module  "I"     Clean
.sp
.fi
or from within an afterstep menu file by 'Clean':
.nf
.sp
% cat "~/GNUstep/Library/AfterStep/start/Modules/Clean"
        Module "Clean" Clean
        MiniPixmap "mini-as.xpm"
.fi

To stop the \fBClean\fP module, a menu item can be used:
.nf
.sp
% cat "~/GNUstep/Library/AfterStep/start/Modules/Clean_Stop"
	KillModuleByName "Stop Clean" Clean
	MiniPixmap "mini-as.xpm"
.fi

.SH DESCRIPTION
The \fBClean\fP module performs actions on windows which have not had
the keyboard focus for a specific length of time. This is intended to
help alleviate the problem of leaving programs running in unused
portions of your desktop.

.SH OPTIONS
.IP "\-f \fIconfig_file\fP"
Use \fIconfig_file\fP instead of the default configuration file.

.IP "\-h | \--help"
Print a usage message and exit successfully.

.IP "\-v | \--version"
Print version information and exit successfully.

.SH INITIALIZATION
During initialization, \fBClean\fP will search for a configuration file
which describes the time-outs and the actions to take.  This file will be 
the one specified by the \fI\-f\fP option (if given).  Else, the 
~/GNUstep/Library/AfterStep/clean file will be used if it exists.  Else, 
/usr/local/share/afterstep/clean will be used.

If the \fBClean\fP executable is linked to another name, ie 'ln -s
\fBClean\fP \fBOtherClean\fP', then another module called \fBOtherClean\fP
can be started, with a completely different set of configurations than
\fBClean\fP, simply by adding \fI*OtherClean\fP options following the same
\fI*Clean\fP configuration options below to the \fIclean\fP file. This way
multiple Cleaning modules can be used.

.SH INVOCATION
\fBClean\fP can be invoked from the \fIautoexec\fP file or by binding the
action "Module Clean \fI[args]\fP" to a key-stroke in a \fIfeel\fP file,
or from a menu item as  shown in the Synopsis section above.

AfterStep will search through the directories specified in the
\fIModulePath\fP configuration in the \fIbase.#bpp\fP (where # is your
colordepth) to attempt to locate Clean.

.SH CONFIGURATION FILE
\fBClean\fP reads the \fIclean\fP file when it starts up and looks for
its configuration options:

.IP "*Clean \fBseconds\fP \fBcommand\fP"
Tells Clean to sleep for x number of seconds after a window loses focus
before it executes the given command; where \fBcommand\fP is any built-in
AfterStep command.

.SH EXAMPLES
At most 3 actions can be specified on any one line.

.IP "*Clean 3600 Iconify 1"
This will sleep for 3600 seconds (1 hour) and then \fIIconify\fP all windows not having been given focus within that 1 hour.  The 1 after \fIIconify\fP defines that only iconification should occur.  Refer to the \fIafterstep.man\fP for further detail on \fIIconify\fP options.

.IP "*Clean 86400 Delete"
This will sleep for 86400 seconds (24 hours) and then \fBDelete\fP all windows not having been given focus within that 24 hours.  Refer to the \fIafterstep.man\fP for further detail on \fBDelete\fP.

.SH FILES
.IP \fI/usr/local/share/afterstep/clean\fP
The global configuration file for system wide defaults of all users.

.IP \fI$HOME/GNUstep/Library/AfterStep/clean\fP
This file should be copied from the \fI/usr/local/share/afterstep/clean\fP and edited to suit a users specific requirements.  This file, if it exists,
will override the system wide default file.

.SH AUTHORS
Robert Nation 
.nf
David Taylor <dtaylor@cit.nepean.uws.edu.au>
.nf
David Mihm <dmihm@swbell.net> [Man page]

.SH SEE ALSO
.BR afterstep (1)


ChangeLog

2003-09-25  sasha

	* generating ChangeLog

2003-09-24  sasha

	* Completed look.Glass ; renamed PointerFore, PointerBack to
	  CursorFore, CursorBack to match old Cursor option; Moved Cursor
	  definitions from feel into look Completed Colorscheme's support
	  for Cursor's color

2003-09-23  sasha

	* Added look options for mouse pointer coloring: PointerFore and
	  PointerBack; Repaired code to update window's decorations when
	  stickiness flag is toggled;

2003-09-22  sasha

	* Added cursors to default feel file; enable cursor switching while
	  hovering over different elements of the interface; Added code to
	  change color of the cursor according to what colorscheme is set
	  to - may need to tweak a bit , or make colors configurable
	* Fixed locale parsing from LANG variable; Fixed module startup to
	  pass on current locae specified on command line

2003-09-19  sasha

	* added debugging code for cursor loading

2003-09-16  sasha

	* fixed bug with windows teleporting back to original position, if
	  moved/resized too fast

2003-09-15  sasha

	* updated configure scripts
	* Fixed configure.in to properly enable i18n in libAfterImage as
	  requested. Internationalized text should be displayed properly
	  now, even if its not in UTF8; Update src/test to work with new
	  libs and added test code for i18n testing
	* Fixed unquoted string in i18N code reported by Alex Mezin

2003-09-14  sasha

	* Added checks for locale and proper initialization of
	  libAfterImage t3ext drawing in accordance to its value; Fixed
	  problem with EdgeScroll being mangled when just look changes;
	  Fixed recently introduced bug causing wierd corners in framned
	  windows upon the look change
	* Fixed typoe for msg type in SendString being declared as some
	  wierd type of unsigned long long  - possibly causing wierdness on
	  64 bit machines

2003-09-13  sasha

	* implemented feedback functionality via sending the e-mail to
	  sasha :)

2003-09-12  sasha

	* added configure option to send e-mail back to developer on
	  installation; Fixed canvas border width handling while resizing
	  and in Pager;

2003-09-11  sasha

	* fixed handling of WM_PROTOCOL changes where WM_COMMAND would get
	  lost and session management would stop working; Found bug with
	  huge impact - all the geometry calculations were off as we were
	  forgetting border width of the frame window; Fixed handling of
	  button presses on border of the frame

2003-09-10  sasha

	* fixed NeXTish colorscheme somewhat
	* Fixed workspace state save/restore to allow terms to exec command
	  using -e, and to have argv's with spaces in them; Fixed frame
	  corner sizing to make it same width as is tbar size;

2003-09-09  sasha

	* More fixing Pager's geometry; Added -i cmd line parameter to
	  ascompose to allow for 'include' files (such as colorschemes,
	  etc.); Updated configure scripts to properly detect headers when
	  its run from libAfterBase; Updated version to beta2; Updated TEAM
	  file

2003-09-08  sasha

	* Fixed Pager to properly account for label size while sizing
	  min-desktop windows

2003-09-07  sasha

	* somewhat debugged Pager's geometry handling - need to add some
	  code to properly recalculate Pager's geometry if its soecified as
	  ratio of the main desktop vs. exact geometry

2003-09-03  sasha

	* added new AfterStep log icon; fixed wharf config file

2003-09-02  sasha

	* added new xml tag : <color> to define symbolic color names
	* changed handling of tiled images to clip it properly when
	  Right/Left align is used

2003-08-30  sasha

	* debugged workspace state handling code, and improved Wait loop

2003-08-29  sasha

	* mostly implemented new workspace state saving and restoring

2003-08-28  sasha

	* fixed workspace status generator to not save xrm params, as noone
	  seems to support thoise
	* added caching of pixmaps for unfocused bas in window decorations
	  to speed up fast focus switching
	* 64-bit fixes: CARD32 instead of Window type
	* 64-bit fixes: properties and strlen warnings
	* 64-bit fixes: converted intermodule communications to always use
	  32bit values
	* more 64 bit fixes

2003-08-27  sasha

	* 64-bit compilation fixes
	* Fixed slowness in intermodule communications manifesting itself
	  when desktops are switched
	* fixed handling of non-default MyBackgrounds, and proper adding of
	  DeskConfigs while switching desks
	* synching
	*  fixed segfaulting in Pager on look switches or after window
	  restacking
	* Fixed handling of hints flags - it was doing nothing before if
	  flags == 0 - was screwing MOtif Hints like totatlly; Fixed shaped
	  window handling in WinList; Added check to not do too much
	  redrawing when WinLIst is receiving window list - may want to do
	  the same in Pager

2003-08-26  sasha

	*  added parameters to import funcions, to allow for sized
	  image/preview stuff; Removed compression buffer from ASImage
	  structure to save up some memory; Hopefully fixed hints handling
	  on 64 bit hw

2003-08-24  sasha

	* fixed event handling when windows is getting mapped - no more
	  lost events due to not selected mask ; Fixed initial status when
	  mouse pointer is out of the screen; Fixed ICCCM size hints
	  handling to have BaseSize and MinSize interchange each other when
	  one is ommited - fixes GIMPs initial placement

2003-08-23  sasha

	* added new 3 icons to complete the set of default icons: switched
	  AfterStep to use ~/.afterstep instead of ridiculous ~/GNUstel/L/A
	  stuff

2003-08-22  sasha

	* Fixed handling of shaped windows - shape is not erroneously
	  cleared, and client's shape is subtracted from decoration's shape
	* fixed autodetection of zipped xpms and added 6 more icons to
	  default iconset

2003-08-21  sasha

	* more improvements to asimbrowser

2003-08-20  sasha

	* fixed makefile for ascp, when fltk is not found; added better
	  Style for BitchX to database; remnoved scrollbar from BitchX in
	  Wharf

src/Clean/Clean.c
/*
 * Copyright (c) 1997 Guylhem Aznar <guylhem@oeil.qc.ca>
 * Copyright (c) 1994 Robert Nation
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#define TRUE 1
#define FALSE
#define IN_MODULE
#define MODULE_X_INTERFACE

#include "../../configure.h"
#ifdef ISC
#include <sys/bsdtypes.h>	/* Saul */
#endif

#include <stdio.h>
#include <signal.h>
#include <fcntl.h>
#include <string.h>
#include <sys/wait.h>
#include <sys/time.h>
#include <time.h>
#include <sys/stat.h>
#if defined ___AIX || defined _AIX || defined __QNX__ || defined ___AIXV3 || defined AIXV3 || defined _SEQUENT_
#include <sys/select.h>
#endif
#include <unistd.h>
#include <ctype.h>
#include <stdlib.h>

#include <X11/Xlib.h>
#include <X11/Xutil.h>

#include "../../include/aftersteplib.h"
#include "../../include/afterstep.h"
#include "../../include/style.h"
#include "../../include/screen.h"
#include "../../include/module.h"

#include "Clean.h"

char *MyName;
Display *dpy;			/* which display are we talking to */
int screen;
ScreenInfo Scr;

int fd_width;
int fd[2];
struct list *list_root = NULL;
unsigned long current_focus = 0;

long period[3] =
{-1, -1, -1};
long maxperiod = -1;
char command[3][256];
int num_commands = 0;

unsigned long next_event_time = 0;
unsigned long t0;

void
version (void)
{
  printf ("%s version %s\n", MyName, VERSION);
  exit (0);
}

void
usage (void)
{
  printf ("Usage:\n"
	  "%s [--version] [--help]\n", MyName);
  exit (0);
}

/***********************************************************************
 *
 *  Procedure:
 *	main - start of module
 *
 ***********************************************************************/
int
main (int argc, char **argv)
{
  int i;
  char *global_config_file = NULL;

  /* Save our program name - for error messages */
  SetMyName (argv[0]);
  i = ProcessModuleArgs (argc, argv, &(global_config_file), NULL, NULL, usage);
  if (i == argc)
    usage ();
  
  /* Dead pipes mean AfterStep died */
  signal (SIGPIPE, DeadPipe);

  ConnectX (&Scr, display_name, 0);
  screen = Scr.screen ;

  /* connect to AfterStep */
  fd[0] = fd[1] = ConnectAfterStep (M_ADD_WINDOW |
						M_CONFIGURE_WINDOW |
						M_DESTROY_WINDOW |
						M_FOCUS_CHANGE |
						M_END_WINDOWLIST);

  /* scan config file for set-up parameters */
  LoadConfig (global_config_file, "clean", ParseOptions);

  if (num_commands == 0)
    {
      fprintf (stderr, "%s: Nothing to do!\n", MyName);
      exit (0);
    }

  fd_width = GetFdWidth ();

  /* Create a list of all windows */
  /* Request a list of all windows,
   * wait for ConfigureWindow packets */
  SendInfo (fd, "Send_WindowList", 0);

  Loop (fd);

  return 0;
}


void
ParseOptions (const char *filename)
{
  FILE *file;
  char *line, *tline;
  int len;

  file = fopen (filename, "r");
  if (file != NULL)
    {
      line = (char *) safemalloc (MAXLINELENGTH);
      len = strlen (MyName);
      while (((tline = fgets (line, MAXLINELENGTH, file)) != NULL) && (num_commands < 3))
	{
	  while (isspace (*tline))
	    tline++;
	  if ((*tline == '*') && (!mystrncasecmp (tline + 1, MyName, len)))
	    {
	      tline += len + 1;
	      sscanf (tline, "%ld", &period[num_commands]);
	      if (period[num_commands] > maxperiod)
		maxperiod = period[num_commands];
	      while (isspace (*tline))
		tline++;
	      while ((!isspace (*tline)) && (*tline))
		tline++;
	      while (isspace (*tline))
		tline++;	/* points to "command" field */
	      strcpy (command[num_commands], tline);
	      num_commands++;
	    }
	}
      fclose (file);
      free (line);
    }
}

/***********************************************************************
 *
 *  Procedure:
 *	Loop - wait for data to process
 *
 ***********************************************************************/
void
Loop (int *fd)
{
  unsigned long header[3], *body;
  fd_set in_fdset;
  struct itimerval value;
  int retval;

  while (1)
    {
      FD_ZERO (&in_fdset);
      FD_SET (fd[1], &in_fdset);

      /* Find out when we have to iconify or whatever a window */
      find_next_event_time ();

      /* set up a time-out, in case no packets arrive before we have to
       * iconify something */
      value.it_value.tv_usec = 0;
      value.it_value.tv_sec = next_event_time - t0;

#ifdef __hpux
      if (value.it_value.tv_sec > 0)
	retval = select (fd_width, (int *) &in_fdset, 0, 0, &value.it_value);
      else
	retval = select (fd_width, (int *) &in_fdset, 0, 0, NULL);
#else
      if (value.it_value.tv_sec > 0)
	retval = select (fd_width, &in_fdset, 0, 0, &value.it_value);
      else
	retval = select (fd_width, &in_fdset, 0, 0, NULL);
#endif
      if (FD_ISSET (fd[1], &in_fdset))
	{
	  /* read a packet */
	  if (ReadASPacket (fd[1], header, &body) > 0)
	    {
	      /* dispense with the new packet */
	      process_message (header[1], body);
	      free (body);
	    }
	}
    }
}


/***********************************************************************
 *
 *  Procedure:
 *	Process message - examines packet types, and takes appropriate action
 *
 ***********************************************************************/
void
process_message (unsigned long type, unsigned long *body)
{
  struct list *l;

  switch (type)
    {
    case M_ADD_WINDOW:
      if (!find_window (body[0]))
	add_window (body[0]);
      break;
    case M_CONFIGURE_WINDOW:
      if (!find_window (body[0]))
	add_window (body[0]);
      break;
    case M_DESTROY_WINDOW:
      remove_window (body[0]);
      break;
    case M_FOCUS_CHANGE:
      l = find_window (body[0]);
      if (l == 0)
	{
	  add_window (body[0]);
	  l = find_window (body[0]);
	  update_focus (l, body[0]);
	}
      else
	update_focus (l, body[0]);
      break;
    default:
      break;
    }
}


/***********************************************************************
 *
 *  Procedure:
 *	find_window - find a window in the current window list 
 *
 ***********************************************************************/
struct list *
find_window (unsigned long id)
{
  struct list *l;

  if (list_root == NULL)
    return NULL;

  for (l = list_root; l != NULL; l = l->next)
    {
      if (l->id == id)
	return l;
    }
  return NULL;
}


/***********************************************************************
 *
 *  Procedure:
 *	remove_window - remove a window in the current window list 
 *
 ***********************************************************************/
void
remove_window (unsigned long id)
{
  struct list *l, *prev;

  if (list_root == NULL)
    return;

  prev = NULL;
  for (l = list_root; l != NULL; l = l->next)
    {
      if (l->id == id)
	{
	  if (prev != NULL)
	    prev->next = l->next;
	  else
	    list_root = l->next;
	  free (l);
	  return;
	}
      prev = l;
    }
}


/***********************************************************************
 *
 *  Procedure:
 *	add_window - add a new window in the current window list 
 *
 ***********************************************************************/
void
add_window (unsigned long new_win)
{
  struct list *t;

  if (new_win == 0)
    return;

  t = (struct list *) safemalloc (sizeof (struct list));
  t->id = new_win;
  t->last_focus_time = time (0);
  t->actions = 0;
  t->next = list_root;
  list_root = t;
}


/***********************************************************************
 *
 *  Procedure:
 *	mark the window which currently has focus, so we don't iconify it
 *
 ***********************************************************************/
void
update_focus (struct list *l, unsigned long win)
{
  struct list *t;

  t = find_window (current_focus);
  if (t != NULL)
    {
      t->last_focus_time = time (0);
      t->actions = 0;
    }

  if (l != NULL)
    {
      l->last_focus_time = time (0);
      l->actions = 0;
    }
  current_focus = win;
}

/***********************************************************************
 *
 *  Procedure:
 *	SIGPIPE handler - SIGPIPE means AfterStep is dying
 *
 ***********************************************************************/
void
DeadPipe (int nonsense)
{
  exit (0);
}

/***********************************************************************
 *
 *  Procedure:
 *	checks to see if we are supposed to take some action now,
 *      finds time for next action to be performed.
 *
 ***********************************************************************/
void
find_next_event_time (void)
{
  struct list *t;

  t0 = time (0);
  next_event_time = t0;

  if (list_root == NULL)
    return;

  next_event_time = t0 + maxperiod;

  for (t = list_root; t != NULL; t = t->next)
    {
      if (t->id != current_focus)
	{
	  if ((period[2] > 0) && (t->last_focus_time + period[2] <= t0))
	    {
	      if (!(t->actions & ACTION1))
		{
		  SendInfo (fd, command[2], t->id);
		  t->actions |= ACTION1;
		}
	    }
	  else if ((period[1] > 0) && (t->last_focus_time + period[1] <= t0))
	    {
	      if (!(t->actions & ACTION2))
		{
		  SendInfo (fd, command[1], t->id);
		  t->actions |= ACTION2;
		}
	    }
	  else if ((period[0] > 0) && (t->last_focus_time + period[0] <= t0))
	    {
	      if (!(t->actions & ACTION3))
		{
		  SendInfo (fd, command[0], t->id);
		  t->actions |= ACTION3;
		}
	    }

	}
      else
	t->last_focus_time = t0;

      if ((period[0] > 0) && (t->last_focus_time + period[0] > t0) &&
	  (t->last_focus_time + period[0] < next_event_time))
	next_event_time = t->last_focus_time + period[0];
      else if ((period[1] > 0) && (t->last_focus_time + period[1] > t0) &
	       (t->last_focus_time + period[1] < next_event_time))
	next_event_time = t->last_focus_time + period[1];
      else if ((period[2] > 0) && (t->last_focus_time + period[2] > t0) &&
	       (t->last_focus_time + period[2] < next_event_time))
	next_event_time = t->last_focus_time + period[2];
    }
}
